File: 2021/day03.py
from functools import reduce


def most_common_bit(nums, pos):
    return int(sum(num[pos] == '1' for num in nums) >= len(nums) / 2)


def least_common_bit(nums, pos):
    return 1 - most_common_bit(nums, pos)


def part1(nums):
    gamma_rate = reduce(lambda acc, pos: (acc << 1) | most_common_bit(nums, pos), range(len(nums[0])))
    print(gamma_rate * (~gamma_rate & ((1 << len(nums[0])) - 1)))


def part2(nums):
    def filter_down(arr, strategy):
        pos = 0
        while len(arr) > 1:
            res = str(strategy(arr, pos))
            arr = list(filter(lambda n: n[pos] == res, arr))
            pos += 1
        return int(arr[0], 2)

    print(filter_down(nums, most_common_bit) * filter_down(nums, least_common_bit))


with open("input/day03.txt") as f:
    v = list(map(lambda l: l.replace('\n', ''), f.readlines()))
    part1(v)
    part2(v)

File: 2021/day03.go
package main

import (
	"aoc/utils"
	"fmt"
)

func mostCommonBit(values []string, pos int) int {
	c := 0
	for _, value := range values {
		if value[pos] == '1' {
			c++
		} else {
			c--
		}
	}
	if c >= 0 {
		return 1
	}
	return 0
}

func leastCommonBit(values []string, pos int) int {
	return 1 - mostCommonBit(values, pos)
}

// Part1 of day3
func Part1(values []string) {
	gammaRate := 0
	for pos := range values[0] {
		gammaRate = (gammaRate << 1) | mostCommonBit(values, pos)
	}
	epsilonRate := ^gammaRate & ((1 << len(values[0])) - 1)
	fmt.Println(gammaRate * epsilonRate)
}

// Part2 of day3
func Part2(values []string) {
	find := func(arr []string, strategy func([]string, int) int) int {
		for pos := 0; len(arr) > 1; pos++ {
			res := strategy(arr, pos)
			arr = utils.Filter(arr, func(value string) bool {
				return int(value[pos]-'0') == res
			})
		}
		return utils.ToIntFromBinaryOrDie(arr[0])
	}
	fmt.Println(find(values, mostCommonBit) * find(values, leastCommonBit))
}

func main() {
	values := utils.ReadFileLinesOrDie("./input/day03.txt")
	Part1(values)
	Part2(values)
}
